### PR TITLE
feat: adding OAuth support to GitHub factory.

### PR DESCRIPTION
**Pull request by cccs-eric**
### What does this PR do?
This PR adds OAuth authentication support for GitHub factory.

### What issues does this PR fix or reference?
Issue [19966](https://github.com/eclipse/che/issues/19966)

### How to test this PR?

1. Make sure you have a linked GitHub account in Keycloak.
2. Create a private git repository on GitHub with a devfile in it
3. Try to open this private repository using the Che factory (whether by specifying the git repository or the path to the devfile)

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubPersonalAccessTokenFetcher.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+      }
+      if (Arrays.binarySearch(tokenScopes, parentScope) < 0
+          && Arrays.binarySearch(tokenScopes, requiredScope) < 0) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private String getLocalAuthenticateUrl() {
```
Comment (mshaposhnik) [inline wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubPersonalAccessTokenFetcher.java:198]: Yes, this URL is passed to the client when OAuth login tourine should be started from scratch

**Comment by mshaposhnik** [comment]
LGTM !

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+import java.util.concurrent.Executors;
+import java.util.function.Function;
+import org.eclipse.che.api.factory.server.scm.exception.ScmBadRequestException;
+import org.eclipse.che.api.factory.server.scm.exception.ScmCommunicationException;
+import org.eclipse.che.api.factory.server.scm.exception.ScmItemNotFoundException;
+import org.eclipse.che.commons.lang.concurrent.LoggingUncaughtExceptionHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** GitHub API operations helper. */
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java:42]: Can we have some tests for this class?

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  private static final Duration DEFAULT_HTTP_TIMEOUT = ofSeconds(10);
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  public GithubApiClient() {
+    this.apiServerUrl = URI.create(GITHUB_API_SERVER);
+    this.scmServerUrl = URI.create(GITHUB_SERVER);
+    this.httpClient =
+        HttpClient.newBuilder()
+            .executor(
+                Executors.newCachedThreadPool(
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java:63]: singleThread?

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java (hunk 2 above)
Comment (cccs-eric) [reply wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java:44]: Added a full test suite.

**Comment by skabashnyuk** [comment]
@cccs-eric may I ask if you sign ECA? https://www.eclipse.org/legal/ECA.php

**Comment by cccs-eric** [comment]
@skabashnyuk Didn't know there was a new version, so I just did:
image

File: wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubAuthorizingFileContentProvider.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  GithubAuthorizingFileContentProvider(
+      GithubUrl githubUrl,
+      URLFetcher urlFetcher,
+      GitCredentialManager gitCredentialManager,
+      PersonalAccessTokenManager personalAccessTokenManager) {
+    super(githubUrl, urlFetcher, personalAccessTokenManager, gitCredentialManager);
+  }
+
+  /**
+   * Formating OAuth token as HTTP Authorization header.
```
Comment (mshaposhnik) [inline wsmaster/che-core-api-factory-github/src/main/java/org/eclipse/che/api/factory/server/github/GithubAuthorizingFileContentProvider.java:31]:
```suggestion
   * Formatting OAuth token as HTTP Authorization header.
```

**Comment by mshaposhnik** [comment]
This is quite good, guys. I think we can merge it in current state. As a product feature, it still have some missing puzzles (like documentation, random files resolver etc) but i think we can manage that by our own forces :- ).

**Review by mshaposhnik — APPROVED** [review]
This is quite good, guys. I think we can merge it in current state. As a product feature, it still have some missing puzzles (like documentation, random files resolver etc) but i think we can manage that by our own forces :- ).

**Comment by cccs-eric** [comment]
Thanks for the review @mshaposhnik.  For my own knowledge, what do you mean by "random files resolver"?

**Comment by mshaposhnik** [comment]
> Thanks for the review @mshaposhnik. For my own knowledge, what do you mean by "random files resolver"?

It's a new endpoint, allowing fetching any files from secured/unsecured repos (since newest devfile versions may require additional files to build an WS). You can look at `ScmService` in main branch

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**